### PR TITLE
[tt-train] Cross entropy backward pass

### PR DESCRIPTION
**Pull request by vmelnykovTT**
### Ticket
[[Feature Request] Cross entropy loss forward and backward functions](https://github.com/tenstorrent/tt-metal/issues/14113)

### Problem description
Since the cross_entropy_bw operation was missing, it has now been implemented. Relevant models were updated to use it, resulting in improved performance.

### What's changed
- Added cross_entropy_bw operation
- Updated cross_entropy_loss implementation in losses.cpp
- Switched mnist, gpt and llama models to use the new cross_entropy_loss
- Added utils files/fucntions fo tt-train internal kernels

### Comparison with composite

Mean Step Time for cross_entropy_bw: 688.39 ms
Mean Step Time for baseline: 785.94 ms

Performance (end-to-end training):
step_time_gpt2s

Loss(first 200 steps for gpt2s):
losses_gpt2s

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/dataflow/writer_cross_entropy_bw_interleaved_start_id.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        cb_wait_front(cb_target_idx, onetile);
+        auto target_indexes_l1_ptr = reinterpret_cast<uint32_t *>(get_read_ptr(cb_target_idx));
+
+        for (uint32_t c = 0, idx = r * Wt; c < Wt; c += block_size) {
+            cb_wait_front(cb_output_idx, block_size);
+            uint32_t l1_read_addr = get_read_ptr(cb_output_idx);
+
+            auto write_ouput_l1_ptr = reinterpret_cast<uint16_t *>(l1_read_addr);
+
+            for (uint32_t h = 0; h < TILE_HEIGHT; ++h) {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/dataflow/writer_cross_entropy_bw_interleaved_start_id.cpp:62]: Discussed offline the way how we can optimize this code.

In this implementation time complexity will be O(Wt * TILE_HEIGHT) vs proposed (TILE_HEIGHT log TILE_HEIGHT + Wt). Feel free to merge it as is, but would be cool to investigate how it affects our performance

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+constexpr uint32_t kNumExpSumAfterReductionTiles = 2U;
+constexpr uint32_t kNumScalerTiles = 1U;  // used it to reduction
+
+constexpr uint32_t kPageElementsNumber = 32U;
+
+const std::string kMaskWDefineKey = "DO_MASK_W";
+const std::string kEverythingFitsInL1DefineKey = "EVERYTHING_FITS_IN_L1";
+
+uint32_t get_block_size(uint32_t num_inner) {
+    const uint32_t max_block_size = 4U;  // 4 is the maximum block size for enabled fp32 dest acc
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp:58]: change `max_block_size` to `3U` (we need one extra register during calculation)

File: tt-train/sources/ttml/metal/ops/common/common_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/common/common_utils.hpp:1]: please rename this file as it is utils file only for reader and writer.

Comment (rfurko-tt) [reply tt-train/sources/ttml/metal/ops/common/common_utils.hpp:1]: two possible solutions:
* three folders (dataflow, compute, program)
* three files (dataflow_utils, compute_utils, program_utils)

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+constexpr uint32_t kNumExpSumBeforeReductionTiles = 2U;
+constexpr uint32_t kNumExpSumAfterReductionTiles = 2U;
+constexpr uint32_t kNumScalerTiles = 1U;  // used it to reduction
+
+constexpr uint32_t kPageElementsNumber = 32U;
+
+const std::string kMaskWDefineKey = "DO_MASK_W";
+const std::string kEverythingFitsInL1DefineKey = "EVERYTHING_FITS_IN_L1";
+
+uint32_t get_block_size(uint32_t num_inner) {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp:57]: change this function to receive `max_block_size` and move it to `common` utilities file (do the same for other operations, thanks in advance)

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+    const uint32_t max_block_size = 4U;  // 4 is the maximum block size for enabled fp32 dest acc
+    for (uint32_t block_size = max_block_size; block_size > 1; block_size--) {
+        if (num_inner % block_size == 0) {  // if num_inner is divisible by block_size - choose this block_size
+            return block_size;
+        }
+    }
+    return 1U;
+}
+
+uint32_t pack_two_bfloat16_to_uint32(float value) {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp:67]: move it to common

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+
+    const uint64_t masks_memory = 2U * kNumMaskTiles * bfloat16_single_tile_size_bytes;
+    const uint64_t scalers_memory = (2U * kNumScalerTiles) * bfloat16_single_tile_size_bytes;  // scaler and matmul
+    const uint64_t output_memory = twice_block_size * bfloat16_single_tile_size_bytes;
+    const uint64_t max_value_memory =
+        (kNumMaxValueAfterReductionTiles + kMaxValueBeforeReductionTiles) * bfloat16_single_tile_size_bytes;
+    const uint64_t exp_sum_memory =
+        (kNumExpSumBeforeReductionTiles + kNumExpSumAfterReductionTiles) * float32_single_tile_size_bytes;
+    const uint64_t input_memory =
+        /* inpunt */ (Wt * bfloat16_single_tile_size_bytes) +
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp:270]: typo in comment

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+    auto cb_exp_sum_after_refuction = create_circular_buffer(
+        program,
+        all_cores,
+        KExpSumAfterReductionCbIndex,
+        precise_data_format,
+        float32_single_tile_size_bytes,
+        kNumExpSumAfterReductionTiles);
+
+    auto cb_mat_mul_reduce =
+        create_circular_buffer(program, all_cores, kMatMulCbIndex, data_format, bfloat16_single_tile_size_bytes, 1U);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/cross_entropy_bw_program_factory.cpp:334]: use constant

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+constexpr auto cb_max_mask = tt::CBIndex::c_3;
+constexpr auto cb_reduction_scaler = tt::CBIndex::c_4;
+constexpr auto cb_max_value_before_reduction = tt::CBIndex::c_5;
+constexpr auto cb_max_value_after_reduction = tt::CBIndex::c_6;
+constexpr auto cb_exp_sum_before_reduction = tt::CBIndex::c_7;
+constexpr auto cb_exp_sum_after_reduction = tt::CBIndex::c_8;
+constexpr auto cb_mat_mul_reduce = tt::CBIndex::c_9;
+constexpr auto cb_output = tt::CBIndex::c_10;
+
+constexpr uint32_t onetile = 1;
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp:50]: 1U

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+                    // it's possible that the padding contains a NaN, and operations like NaN + (-inf) = NaN,
+                    // instead of the expected -inf. similarly, -inf * 0 = NaN.
+                    copy_tile_init(cb_mask);
+                    copy_tile(cb_mask, /* tile_idx */ 0, /* register idx */ mask_register);
+                    mask_tile_init();
+                    mask_tile(working_register, mask_register);  // mask should be next to tile register.
+
+                    copy_tile_init(cb_max_mask);
+                    copy_tile(cb_max_mask, /* tile_idx */ 0, /* register idx */ mask_register);
+
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp:91]: remove empty line or add one between copy_tile and mask_tile

**Review by rfurko-tt — COMMENTED** [review]
Overall looks great, needs a little bit of cleaning, but very close to merging. Great work!

File: tt-train/sources/ttml/metal/ops/common/common_utils.hpp (hunk 3 above)
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/common/common_utils.hpp:1]: I chose the second one approach

File: tt-train/sources/ttml/metal/ops/common/program_utils.hpp [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <bit>
+#include <cstdint>
+
+inline uint32_t get_block_size(uint32_t num_inner, const uint32_t max_block_size = 4U) {
+    for (uint32_t block_size = max_block_size; block_size > 1; block_size--) {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/common/program_utils.hpp:11]: 1U

File: tt-train/sources/ttml/metal/ops/common/program_utils.hpp [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+            return block_size;
+        }
+    }
+    return 1U;
+}
+
+inline uint32_t pack_two_bfloat16_to_uint32(float value) {
+    uint32_t uint32_data = std::bit_cast<uint32_t>(value);
+    uint32_t casted_uint16_data = uint32_data >> 16U;
+    return casted_uint16_data | (casted_uint16_data << 16);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/common/program_utils.hpp:22]: 16U

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/dataflow/reader_cross_entropy_bw_interleaved_start_id.cpp [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+#ifndef EVERYTHING_FITS_IN_L1
+        // read input buffer by blocks to calculate sum(exp(x - max(x))) in row
+        read_block_tiles(cb_input_idx, input_address_generator, Wt, block_size, tile_bytes, idx);
+
+        // read input buffer by blocks to calculate softmax in row
+        read_block_tiles(cb_input_idx, input_address_generator, Wt, block_size, tile_bytes, idx);
+#endif
+
+        // #ifdef EVERYTHING_FITS_IN_L1
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/dataflow/reader_cross_entropy_bw_interleaved_start_id.cpp:119]: remove commented code

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp [hunk 13]
```diff
@@ -247,7 +238,7 @@ CrossEntropyForwardProgramFactory::cached_program_t CrossEntropyForwardProgramFa
     uint32_t mask_w = num_inner % tt::constants::TILE_WIDTH;  // width index of first trash value in tile
 
     // compile arguments
-    uint32_t block_size = get_block_size(Wt);
+    uint32_t block_size = get_block_size(Wt, 4U);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp:241]: please check whether you can set 4 here (based on compute kernel usage of additional registers)

File: tt-train/sources/ttml/metal/ops/rmsnorm_fw/device/rmsnorm_fw_program_factory.cpp [hunk 14]
```diff
@@ -236,7 +221,7 @@ RMSNormForwardProgramFactory::cached_program_t RMSNormForwardProgramFactory::cre
     uint32_t packed_scaler = pack_two_bfloat16_to_uint32(1.F / static_cast<float>(num_inner));
     uint32_t packed_eps = pack_two_bfloat16_to_uint32(args.epsilon);
     uint32_t mask_w = num_inner % tt::constants::TILE_WIDTH;
-    uint32_t block_size = get_block_size(Wt);
+    uint32_t block_size = get_block_size(Wt, 4U);  // TODO: ask Roman to check whtether max_block_size should be 4U
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/rmsnorm_fw/device/rmsnorm_fw_program_factory.cpp:224]: i think it should be 3U

File: tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp [hunk 15]
```diff
@@ -0,0 +1,453 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <compute_kernel_api/cb_api.h>
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp:5]: remove this includes as they are copies of includes below

Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/cross_entropy_bw/device/kernels/compute/cross_entropy_bw_kernel.cpp:5]: Without these files kernel doesn't compile

**Review by rfurko-tt — APPROVED** [review]
LGTM, great work

Do not forget to uncomment `enable_wandb` line before merge.

@dmakoviichuk-tt PTAL

**Review by dmakoviichuk-tt — APPROVED** [review]
good for me but I got an idea how to easily distinguish between kernel and host code.
